10/calculator_test.go

```diff
@@ -0,0 +1,101 @@
+package calculator_test
+
+import (
+	"testing"
+	"calculator"
+)
+
+func TestAdd(t *testing.T) {
+	testCases := []struct {
+		a, b int
+		want int
+	}{
+		{a: 1, b: 1, want: 2},
+		{a: 2, b: 2, want: 4},
+		{a: 6, b: 7, want: 13},
+		{a: -6, b: -7, want: -13},
+		{a: 6, b: -7, want: -1},
+		{a: 0, b: -7, want: -7},
+		{a: 0, b: 0, want: 0},
+		{a: -1, b: 0, want: -1},
+	}
+	for _, testCase := range testCases {
+		got := calculator.Add(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Add(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
+		}
+	}
+}
+
+func TestSubtract(t *testing.T) {
+	testCases := []struct {
+		a, b int
+		want int
+	}{
+		{a: 1, b: 1, want: 0},
+		{a: 2, b: 1, want: 1},
+		{a: 6, b: 7, want: -1},
+		{a: 7, b: 6, want: 1},
+		{a: -6, b: -7, want: 1},
+		{a: 6, b: -7, want: 13},
+		{a: 0, b: -7, want: 7},
+		{a: 0, b: 0, want: 0},
+		{a: -1, b: 0, want: -1},
+		{a: -1, b: 7, want: -8},
+	}
+	for _, testCase := range testCases {
+		got := calculator.Subtract(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Subtract(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
+		}
+	}
+}
+
+func TestMultiply(t *testing.T) {
+	testCases := []struct {
+		a, b int
+		want int
+	}{
+		{a: 1, b: 1, want: 1},
+		{a: 2, b: 1, want: 2},
+		{a: 6, b: 7, want: 42},
+		{a: -6, b: -7, want: 42},
+		{a: 6, b: -7, want: -42},
+		{a: 0, b: -7, want: 0},
+		{a: 0, b: 0, want: 0},
+		{a: -1, b: 0, want: 0},
+		{a: -1, b: 7, want: -7},
+	}
+	for _, testCase := range testCases {
+		got := calculator.Multiply(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Multiply(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
+		}
+	}
+}
+
+func TestDivide(t *testing.T) {
+	testCases := []struct {
+		a, b        int
+		want        int
+		errExpected bool
+	}{
+		{a: 1, b: 1, want: 1, errExpected: false},
+		{a: 2, b: 1, want: 2, errExpected: false},
+		{a: 6, b: -7, want: 0, errExpected: false},
+		{a: -1, b: 7, want: 0, errExpected: false},
+		{a: 10, b: 0, want: 0, errExpected: true},
+	}
+	for _, testCase := range testCases {
+		got, _ := calculator.Divide(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Divide(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
+		}
+
+		_, gotErr := calculator.Divide(testCase.a, testCase.b)
+		if testCase.errExpected == true && gotErr == nil {
```

Comment:
This covers one of the four possibilities, but we also need to catch the case where `errExpected == false` and we _did_ get an error, don't we? So a long way to write this would be:

```go
if (testCase.errExpected && gotErr == nil) || (!testCase.errExpected && gotErr != nil) {
```

Which is a little indigestible. A little Boolean algebra (finally, some use for CS class!) gives us this:

```go
if testCase.errExpected != (gotErr != nil) {
```

In other words, the truth or falsehood of the statement "We got an error" should be the same as the truth or falsehood of "We expected an error". If it's not, the test should fail.